services/accumulo-utils/src/main/java/datawave/accumulo/util/security/UserAuthFunctions.java

```diff
@@ -0,0 +1,146 @@
+     * query while also preventing privilege escalation.
+     * <p>
+     * If {@code requestedAuths} contains elements that don't appear in {@link DatawaveUser#getAuths()}, implementors may opt to throw a runtime exception
+     * indicating that authorizations have been requested which the user has not been granted.
+     * <p>
+     * Additionally, implementors should return a default {@link Authorizations} instance (i.e., {@code new Authorizations()}) rather than {@code null} to
+     * denote that the request is invalid based on the given inputs.
+     *
+     * @param requestedAuths
+     *            The set of requested Accumulo authorizations (comma-delimited)
+     * @param user
+     *            The DataWave user for whom authorizations are being requested
+     *
+     * @return {@link Authorizations} instance appropriate for the given inputs
+     */
+    Authorizations getRequestedAuthorizations(String requestedAuths, DatawaveUser user);
+    
+    /**
+     * This method provides a merged view of the auths from the specified proxy chain and from those in {@code primaryUserAuths}, e.g., as precomputed by
+     * {@link #getRequestedAuthorizations(String, DatawaveUser)} perhaps.
+     *
+     * <p>
+     * The resulting set may be applied by DataWave at scan time to ensure that any data returned from Accumulo is suitable for any/all users in the given chain
+     *
+     * @param primaryUserAuths
+     *            Authorizations associated with the "primary" user, i.e., the initiating user to whom scan results will ultimately be delivered
+     * @param proxyChain
+     *            Collection of proxies denoting the complete user chain for a given request
+     *
+     * @return A set of {@link Authorizations}, one per user entity represented by the user chain including {@code primaryUserAuths}
+     */
+    LinkedHashSet<Authorizations> mergeAuthorizations(Authorizations primaryUserAuths, Collection<? extends DatawaveUser> proxyChain);
+    
+    /**
+     * Default implementation of {@link UserAuthFunctions}
+     */
+    class Default implements UserAuthFunctions {
+        
+        /**
+         *
+         * @return IFF every element in {@code requestedAuths} also exists in {@link DatawaveUser#getAuths()} then {@code requestedAuths} is translated to
+         *         {@link Authorizations} and returned. If either of {@code requestedAuths} or {@link DatawaveUser#getAuths()} is empty/null, then
+         *         {@link Authorizations#EMPTY} is returned. If any requested auths are missing in {@link DatawaveUser#getAuths()}, then an exception is thrown
+         *
+         * @throws IllegalArgumentException
+         *             if {@code requestedAuths} contains elements that do not exist in {@link DatawaveUser#getAuths()}
+         */
+        @Override
+        public Authorizations getRequestedAuthorizations(String requestedAuths, DatawaveUser user) {
```

Comment:
The only reason DatawaveUser (and therefore authorization api) needs to be a dependency here is to call getAuths. I'm not sure if this makes sense or not, but one thought is to do the api in terms of java 8 functions or suppliers. For example:
```java
public Authorizations getRequestedAuthorizations(String requestedAuths, Supplier<Collection<String>> authSupplier) {
   Collection<String> auth = authSupplier.get();
  // do stuff with the auths
}
```
then you would call with
```java
DatawaveUser user;
getRequestedAuthorizations("foo, bar", user::getAuths)
```
That won't work for nulls, though, so you'd have to do something like:
```java
public <T> Authorizations getRequestedAuthorizations(String requestedAuths, T authSupplier, Function<T,Collection<String>> authFunction) {
  if (authSupplier != null) {
    Collection<String> auth = authFunction.apply(authSupplier);
  }
```
and then you'd call with
```java
DatawaveUser user;
getRequestedAuthorizations("foo, bar", user, DatawaveUser::getAuths)
```
Just a thought. Like I said, not sure if it's worth it or not.

Reply:
I like the approach. And on one hand, I agree that it's kinda tempting to remove the DatawaveUser dependency in this particular case. I feel like the main argument for keeping it though is that it leaves the door open for implementations to make their own user-dependent transformation decisions, which may not depend on the auths collection alone.

How do you feel about specifying both options in the interface? Then our "Default" impl could have the DatawaveUser version of getRequestedAuthorizations simply delegate to the "T authSupplier" version of the method (?)

Reply:
Or perhaps provide the "T authSupplier" version as a public static method on the interface? To include a ```..., boolean throwOnMissingAuths)``` arg to allow user to toggle between those two use cases as desired? That is, ```false``` would return the intersection, no matter what was in requestedAuths, while ```true``` would yield our default behavior

Reply:
Yeah, it's a toss up. I'm not sure there's much value to having the generic variants if we're going to keep the authorization-api dependency for DatawaveUser. @tomnelson, any thoughts?

Reply:
Yeah, I'm really ok either way, as long as users have some way to easily override our default behavior, if needed

Reply:
@brianloss  Made a few tweaks based on your feedback, while leaving the DatawaveUser in place for now. Added...
```
static Authorizations getRequestedAuthorizations(String requestedAuths, Supplier<Collection<String>> authSupplier, boolean throwOnMissingAuths)
```